### PR TITLE
Export basic ics

### PR DESCRIPTION
**Pull request by Wardormeur**
First step for https://github.com/CoderDojo/community-platform/issues/1246

### CODE REVIEW

**Comment by Wardormeur** [comment]
To be done:
Use user-provided TZ (until we have the event's one)
Make it work with list of events (which is dodgy with pagination, but that's an edge case)

**Comment by niccokunzmann** [comment]
Question: Why is the time zone important? What is the data you have saved about the dojo? Is that date and local time only, like "2nd of march, 14:00 PM"? ICS requires a time zone, true.

**Comment by Wardormeur** [comment]
Information regarding the Dojo are not available at that stage of the service, this service only cares about events :)
In both scenarios anyway, we save dates as UTCZ, which is to say "as the user inputted it". The reason for that is historical : we don't have the tz associated with the location of the Dojo, nor the event (which can vary from the Dojo location) which would allow us to recalculate the time in case the date would in the future change (think DST for example) and because we assume that the user is based on the same TZ than the creator, which is fine in 99% of the case.
TZ could be backfilled based on the location of the event, but a backfill job is a bit more expensive (Did one 2 weeks ago regarding geolocation) for an experiment that could be using the user's TZ
Does this.. more or less answer the question?

**Comment by niccokunzmann** [comment]
This answers it.

File: test/integration/events.load.spec.js [hunk 1]
```diff
@@ -56,5 +56,31 @@ describe('integration:events:load', () => {
...
+      expect(attributes[3]).to.equal('PRODID:adamgibbons/ics');
+      expect(attributes[4]).to.equal('METHOD:PUBLISH');
+      expect(attributes[5]).to.equal('X-PUBLISHED-TTL:PT1H');
+      expect(attributes[6]).to.equal('BEGIN:VEVENT');
+      expect(attributes[7]).to.equal('UID:a60dc59d-2db2-4d5d-a6d3-c08473dee5d4');
+      expect(attributes[8]).to.equal('SUMMARY:Test event 1');
+      expect(attributes[9]).to.match(/DTSTAMP:[0-9]+T[0-9]+Z/);
+      expect(attributes[10]).to.match(/DTSTART:[0-9]+T[0-9]+Z/);
+      expect(attributes[11]).to.match(/DTEND:[0-9]+T[0-9]+Z/);
+      expect(attributes[12]).to.equal('URL:https://zen.coderdojo.com/events/a60dc59d-2db2-4d5d-a6d3-c08473dee5d4');
```
Comment (niccokunzmann) [inline test/integration/events.load.spec.js]: nice! This is a feature I like to have.

File: test/integration/events.spec.js [hunk 2]
```diff
@@ -168,6 +168,69 @@ describe('integration:events', () => {
...
+      expect(attributes[29]).to.match(/DTEND:[0-9]+T[0-9]+Z/);
+      expect(attributes[30]).to.equal('URL:https://zen.coderdojo.com/events/84c0310e-49ff-4607-99da-a5abb9fb5641');
+      expect(attributes[31]).to.equal('ORGANIZER;CN=CoderDojo:mailto:info@coderdojo.com');
+      expect(attributes[32]).to.equal('END:VEVENT');
+      expect(attributes[33]).to.equal('BEGIN:VEVENT');
+      expect(attributes[34]).to.equal('UID:bcef18f8-b5ff-43a9-bc2c-7109f6e5dc20');
+      expect(attributes[35]).to.equal('SUMMARY:Test event 6');
+      expect(attributes[36]).to.match(/DTSTAMP:[0-9]+T[0-9]+Z/);
+      expect(attributes[37]).to.match(/DTSTART:[0-9]+T[0-9]+Z/);
+      expect(attributes[38]).to.match(/DTEND:[0-9]+T[0-9]+Z/);
```
Comment (niccokunzmann) [inline test/integration/events.spec.js]: Hm. This is `attributes[NUMBER]` can become a problem in later stages of development when attributes are removed. This makes it harder to re-arrange them. Maybe there is something like

```
expect.any(attributes).to.match(/DTEND:[0-9]+T[0-9]+Z/);
expect.one(attributes).to.match(/DTEND:[0-9]+T[0-9]+Z/);
```

File: test/unit/events/formatter.spec.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  });
+  describe('ICS', () => {
+    it('should format a single event into an ics', () => {
+      const result = {
+        id: 1,
+        startTime: new Date(),
+        endTime: new Date(),
+      };
+      const res = formatter.toICS(result);
+      expect(res).to.have.keys(['productId', 'method', 'uid', 'timestamp', 'start', 'organizer', 'end', 'url']);
```
Comment (niccokunzmann) [inline test/unit/events/formatter.spec.js]: This is testing the keys here. Like above, one could supply a mapping.

File: test/integration/events.spec.js (hunk 2 above)
Comment (niccokunzmann) [reply test/integration/events.spec.js]: Or better ^^

```
expect(attributes[i++]).to.match(/DTEND:[0-9]+T[0-9]+Z/);
```

File: events/formatter.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    // TODO: Add applications if they are loaded
+    // TODO: Add status of the even depending on the applications
+    url: `https://zen.coderdojo.com/events/${e.id}`,
+  }));
+  return ics.createEvents(icsEvents).value;
+}
+module.exports.format = function format(req, res) {
+  let response;
+  res.format({
+    'text/calendar': () => {
```
Comment (niccokunzmann) [inline events/formatter.js]: Where is the dispatch taking place? Will a single GET to the dojo url with an ".ics" at the end suffice or do I need to send headers?
I would prefer the .ics ending.

File: test/integration/events.load.spec.js (hunk 1 above)
Comment (Wardormeur) [reply test/integration/events.load.spec.js]: Actually, that's wrong.
I have non-pushed changes, but the RFC is designed to take the URL as an updatable link to the ICS of the event, not an event page. 
I'll have to add the actual link to the description of the event as a .. workaround.

File: test/integration/events.spec.js (hunk 2 above)
Comment (Wardormeur) [reply test/integration/events.spec.js]: I agree it's quite messy once you start having some events. I like the incremental version as a solution.
I think i'll double that with a function to check a single event, and pass the offset based upon the number of properties being checked

File: events/formatter.js (hunk 4 above)
Comment (Wardormeur) [reply events/formatter.js]: Another PR I'll push soon. The events-service is back-office only, the accessible url (from cp-zen-platform) won't require the header but will indeed end in `.ics` as an optional formatting (which means the results from the json and the ics will be consistent without code duplication)

**Pull request merged**